base/src/protocol_state.rs

```diff
@@ -49,10 +76,20 @@ pub struct ProtocolStateBody {
 
 pub trait Header {
     fn get_height(&self) -> Length;
+    fn min_window_density(&self) -> Length;
+    fn sub_window_densities(&self) -> &Vec<Length>;
```

Comment:
Excellent design put `min_window_density` and `sub_window_densities` in the trait 🚀